### ISSUE TITLE
Model versioning

### ISSUE THREAD

**Issue by sstone1** [accordproject/concerto#426]
## Introduction

This is a feature proposal for introducing explicit model versioning into the Concerto modelling language. Today, it is only possible to implement model versioning by convention, by inserting a version number into the namespace of a model. This makes it hard to understand what model versions exist in a set of models and how those model versions depend on each other.

We believe in the use of semantic versioning for models, which would mean that any instance created with a 1.x.y version (for example, 1.0.0) model should be compatible and validate against the latest 1.x.x version (for example, 1.2.3) model. 

```
namespace org.example@1.0.0

concept Person {
  o String name
}
```

A given model manager would be able to load multiple versions of a model, so it could also load a new major version of that model with an incompatible change:

```
namespace org.example@2.0.0

concept Person {
  o String name
  o Integer age 
}
```

## Namespaces
Namespaces are versioned, which implicitly versions all the declarations within the namespace.

Version numbers are composed of:
1. Major number
2. Minor number
3. Patch number
4. Pre-release identifier

Version numbers are encoded within namespaces using the following scheme:

```
namespace org.acme@1.2.3-alpha.0
```

## Imports
A model may reference an explicit version of a namespace - and import types from that namespace. A model must only reference a single version of a namespace.

An import that does not specify a version identifier uses the latest “stable” version of a namespace.

```
import org.acme.Customer@1.2.3-alpha.0 // pre-release
import org.acme.Product // uses "stable"
import org.acme.Product@2.3.4 from https://acme.org/models/product.cto // stable, loaded from a URL
import org.acme.Customer@unstable // use latest prerelease
```

If a version number is specified if must be explicit - i.e. version ranges are not supplied (we do not have a “lock file” concept).

## Instances
JSON serialisation encodes the version of a namespace used for a type. Validation checks that the instance conforms to the version of the namespace. An instance can be verified against the latest (compatible) version of a namespace. On save an instance may optionally be updated to reference the latest compatible version.

An instance would be able to refer to a model version using the model version number that the application used to create that instance:

```
[
  {
    "$class": "org.example.Person@1.0.0",
    "name": "Simon"
  },
  {
    "$class": "org.example.Person@2.0.0",
    "name": "Simon",
    "age": 99
  }
]
```

## Model Manager
Model Manager manages a set of (versioned) namespaces. A model manager may contain version N as well as version N+1 of a namespace, as well as pre-release versions of the namespace.

## Migration
Migration functions may be registered to migrate instances from version N to version N+1 of a namespace.

**Comment by dselman** [comment]
I've started working on this. 

The first step is to extend the CTO grammar to support specifying a version number for a namespace. The least disruptive change is to add a new `VersionedNamespace` grammar rule and modify the `Model` grammar rule to use it 

Eg.

```
ModelVersion 
  = major:DecimalIntegerLiteral '.' minor:DecimalIntegerLiteral '.' patch:DecimalIntegerLiteral {
    return `${decimalIntegerLiteralToString(major)}.${decimalIntegerLiteralToString(minor)}.${decimalIntegerLiteralToString(patch)}`
  }

Namespace
  = NamespaceToken __ ns:QualifiedName ! '@' __ {
  	return ns;
  }

VersionedNamespace
  = NamespaceToken __ ns:QualifiedName '@' version:ModelVersion __ {
  	return `${ns}@${version}`;
  }

NamespaceDeclaration
	= Namespace /
      VersionedNamespace

Model
  = version:Version? ns:NamespaceDeclaration imports:Imports? body:SourceElements? {
      const result = {
        $class: "concerto.metamodel.Model",
        namespace: ns,
        imports: optionalList(imports),
        declarations: optionalList(body)
      };
      if (version) {
        result.concertoVersion = version;
      }
      return result;
    }
```

Note that this approach uses the parser to bake the version number into the namespace, which has some advantages and disadvantages:

- PRO: Do not need to modify the metamodel
- PRO: All existing code which uses the namespace as a key to index models continues to work (i.e. we don't have to change the identifier for a model to be `namespace + version` which I fear would be very impactful.
- PRO: Do not need to modify the metamodel -> CTO text generation
- CON: have to parse the namespace of a model to extract the version number. We can add a utility function for this?
- PRO/CON: instance serialization should work as-is, but will look like:
   ```
   {
      "$class" : "org.acme@1.23.Customer",
      "name" : "Dan"
   }
   ```

Note that we still need to update the grammar rules for `import` to support major version imports. The code that resolves a type from an import declaration will have to be updated to be version aware: https://github.com/accordproject/concerto/blob/master/packages/concerto-metamodel/lib/metamodelutil.js#L228 — looking for the latest version of a specified major version number.

We will have to decide what we do in the validation code with explicit version numbers. For example, we could use the latest compatible version of a model for validation.

We will have to decide what we do when we generate JSON. For example, we could use the latest compatible version number of a model when we serialise an instance to JSON.

**Comment by mttrbrts** [comment]
> PRO: All existing code which uses the namespace as a key to index models continues to work (i.e. we don't have to change the identifier for a model to be namespace + version which I fear would be very impactful.

Can you say a little more about this? As I understand it, this is exactly the motivation for this issue. I want to discriminate between versions and to take advantage of this, I would expect to have to update client code. What's the blast radius within AP?

**Comment by mttrbrts** [comment]
In the CTO Grammar, are you excluding pre-release versions deliberately?

**Comment by dselman** [comment]
> In the CTO Grammar, are you excluding pre-release versions deliberately?

No. I've updated the description to include pre-releases.

**Comment by dselman** [comment]
> > CON: have to decide what we do if we have org.acme and org.acme@2.3.4 in the model manager. Which is the "latest"?
> 
> Do we really want to support this mixed unversioned & versioned model story? Can we push unversioned models into some kind of "deprecated compatibility mode" and drop it in a later major version of Concerto?

Yes, I think we could print some sort of warning for these and add a "strict" flag to the model manager to reject unversioned models. I think unversioned models would have to be considered last - i.e. they are conceptually "below" version 0.0.0, but are considered "stable".
